### ISSUE TITLE
Feature Request: Pass in a function for the baseUrl

### ISSUE THREAD

**Issue by rwilliams3088** [reduxjs/redux-toolkit#4592]
I am in the process of introducing [tsyringe](https://github.com/microsoft/tsyringe) into some internal projects. Dependency Injection is a beautiful thing! I'd like to combine it with RTK Query by allowing an application to configure a given service with a configuration object, and then use that configuration object to control things like the `baseUrl`.

Whilst I could go through the troublesome process of creating my own `baseQuery` function, for the most part I find the provided `fetchBaseQuery` sufficient. And, of course, it's well tested by everyone so I don't need to worry about writing a bunch of tests for it. That said, the limitation I'm running into is that the `baseUrl` essentially has to be hard-coded (though you can use an environment variable or the like that gets configured via your bundler). The tsyringe container is getting initialized only after `createApi` is executed; despite my attempts at re-ordering the imports and such (possibly because I am referencing Symbols and interfaces from the module when configuring the container). 

As such, it would be nice if I could supply a function to determine the `baseUrl`. This could also be useful in other scenarios for dynamically modifying the `baseUrl` if configuration changes, without the need to refresh the page or even recompile and redeploy the application. 

I do have a work-around in the meantime. It looks like you can pass in `undefined` for the baseUrl - in which case it will use the url from the specific endpoint as the whole url. And the endpoint url can be computed dynamically! So this isn't holding anything up, I just think it would be a nice and easy improvement to the API. Thanks!

-----------------------
If you are curious, this is what my work-around looks like:

```
function formatUrl(relativeUrl: string) {
  const config = container.resolve<Readonly<FeatureFlagsServiceConfig>>(FeatureFlagsServiceConfigSymbol);
  const delimiter = relativeUrl.startsWith("/") ? "" : "/";
  return `${config.restAuthority}${delimiter}${relativeUrl}`;
}

export const RestApi = createApi({
  reducerPath: 'feature-flags-rest-api',
  baseQuery: fetchBaseQuery({ baseUrl: undefined }), // NOTE: Each endpoint must provde an absolute url
  endpoints: builder => ({
    getFeatureFlag: builder.query<GetFeatureFlag, GetFeatureFlagVariables>({
      query: (params) => ({
        url: formatUrl(`/feature_flags/${params.application}/${params.feature}`),
        method: "GET",
      }),

      transformResponse: transformResponse(deserializeGetFeatureFlag),
      transformErrorResponse: transformErrorResponse("Feature-Flags", "getFeatureFlag"),
    }),
    ...
  })
});
```

**Comment by phryneas** [comment]
> in which case it will use the url from the specific endpoint as the whole url.

No need to pass `undefined`, it will always do that if your endpoint returns an absolute url instead of a relative one.

---

Generally, you don't need to write your own `fetchBaseQuery`, and we also don't need to add this as a new feature:

Wrap the original `fetchBaseQuery` - it's meant to be composed.

```js
function composeBaseQuery(options){
  return function composedBaseQuery(arg, api){
    const baseUrl = ""; // do what you need here
    return fetchBaseQuery({ ...options, baseUrl })(arg, api)
  }
}
```

**Comment by rwilliams3088** [comment]
@phryneas that doesn't solve the problem. No matter how many composition layers I add, if it has to resolve the `baseUrl` prior to the container being initialized then it doesn't solve the problem. Determining the `baseUrl` must be deferred.

**Comment by phryneas** [comment]
@rwilliams3088 in my example, it is deferred. Please re-read the code.
